# Sneaker Faqtory Tables

This document contains the code used to generate the payment, store, discount, and product tables of the Sneaker Faqtory dataset.

## Payment Table

In [1]:
import random
import numpy as np

paymentmethods = ["Mastercard", "iDeal", "Visa", "PayPal", "Klarna"]

#Generate random payment method data. k = the number of orders
#paymentdata = random.choices(paymentmethods, weights=(40, 10, 8, 24, 18), k=10)

#print(paymentdata)

### Transforming to dataframe in Pandas

In [2]:
import pandas as pd
paymentdf = pd.DataFrame (paymentmethods, columns = ['Payment_Method'])
paymentdf.head()

Payment_Method
0     Mastercard
1          iDeal
2           Visa
3         PayPal
4         Klarna

In [3]:
#paymentdf.index += 1

### Add new column with payment terms for every payment method

In [4]:
#Create list of conditions
conditions = [
    (paymentdf['Payment_Method'] == 'Mastercard'),
    (paymentdf['Payment_Method'] == 'iDeal'),
    (paymentdf['Payment_Method'] == 'Visa'),
    (paymentdf['Payment_Method'] == 'PayPal'),
    (paymentdf['Payment_Method'] == 'Klarna')
    ]

#Create list of values that will be assigned to the conditions
values = ['End of month', 'Debit', 'End of month', 'Debit', 'Net 14']

#Create new column with values
paymentdf['Payment_Terms'] = np.select(conditions, values)

paymentdf.head(10)

Payment_Method Payment_Terms
0     Mastercard  End of month
1          iDeal         Debit
2           Visa  End of month
3         PayPal         Debit
4         Klarna        Net 14

In [5]:
paymentdf.to_csv('Payments.csv')

## Store Table

In [6]:
stores = [['Amsterdam','Oudezijds Voorburgwal', 'Netherlands'],
['Stockholm','Drottninggatan', 'Sweden'],
['London','Carnaby Street', 'United Kingdom'],
['Warsaw','Nowy Świat', 'Poland'],
['Rome','Via del Corso', 'Italy'],
['Madrid','Calle Preciados', 'Spain'],
['Webshop','http://www.sneaker-factory.com']
         ]


#Generate random store data, k = the number of orders
#storedata = random.choices(stores, weights=(18, 8, 16, 5, 11, 7, 35), k=10)

#print(storedata)

In [7]:
storedf = pd.DataFrame (stores, columns = ['Store_Location', 'Address', 'Country'])

storedf.head(10)

Store_Location                         Address         Country
0      Amsterdam           Oudezijds Voorburgwal     Netherlands
1      Stockholm                  Drottninggatan          Sweden
2         London                  Carnaby Street  United Kingdom
3         Warsaw                      Nowy Świat          Poland
4           Rome                   Via del Corso           Italy
5         Madrid                 Calle Preciados           Spain
6        Webshop  http://www.sneaker-factory.com            None

In [35]:
#storedf.index += 1
storedf.index
storedf.head(10)

Store_Location                         Address         Country
0      Amsterdam           Oudezijds Voorburgwal     Netherlands
1      Stockholm                  Drottninggatan          Sweden
2         London                  Carnaby Street  United Kingdom
3         Warsaw                      Nowy Świat          Poland
4           Rome                   Via del Corso           Italy
5         Madrid                 Calle Preciados           Spain
6        Webshop  http://www.sneaker-factory.com            None

In [36]:
storedf.to_csv('Stores.csv')

## Discount Table

In [37]:
# import datetime
# from faker import Faker

# #Generating random dates to test 
# faker = Faker()
# discountlist = [] 
# for _ in range(20):
#     discountlist.append((faker.date_between(start_date='-18y', end_date='now')).strftime("%d %b"))
    
# print(discountlist)

In [38]:
# for i in discountlist:
#     if int(i[:2]) <= 7 and str(i[-3:]) == "Jan":
#         print("New Year discount")
#     elif i[-3:] == "Dec":
#         print("Christmas discount")
#     elif (int(i[:2]) >= 21 and str(i[-3:]) == "Mar") or (int(i[:2]) <= 3 and str(i[-3:]) == "Apr"):
#         print("Spring deal")
#     elif int(i[:2]) == 22 and str(i[-3:]) == "Apr":
#         print("Earth day")
#     else:
#         print("No discount")

In [39]:
discount = [['Christmas', '20 Dec', '24 Dec', 15, 0],
            ['New Year', '02 Jan', '05 Jan', 0, 20],
            ['Earth Day', '22 Apr', '22 Apr', 10, 0],
            ['Spring deal', '18 Mar', '22 Mar', 10, 0]
           ]


## Product table

To create a list of all products with unique ID's, a nested list was created including all possible shoe sizes, colors, and product names/categories. Then, all possible combinations are calculated and put into a dataframe. Each row represents a unique product.


In [59]:
import itertools
import pandas as pd

size = []
for i in range(36,49):
    size.append(i)

color = ['Crimson', 'Navy', 'Olive', 'Baby Blue',
         'Cyan', 'Coral', 'Azure', 'Denim', 'Peach'
        ]

#A nested list with all product specifications is used to find all the possible permutations 
productcategories = [['SWEATZY_MEN','SWEATZY_WOMEN',
               'WOVE_MEN','WOVE_WOMEN',
               'BOOTWEAR_MEN','BOOTWEAR_WOMEN',
               'FROVE_MEN','FROVE_WOMEN',
               'SWOONIX_MEN','SWOONIX_WOMEN'
              ],
               ['Crimson', 'Navy', 'Olive', 'Baby Blue',
                'Cyan', 'Coral', 'Azure', 'Denim', 'Peach'],
               size
              ]

### Listing all permutations for products 

In [60]:
#All permutations are put into a single list and then transformed into a dataframe.
products = list(itertools.product(*productcategories))

productsdf = pd.DataFrame (products, columns = ['Product_Name', 'Product_Color', 'Product_Size'])
productsdf.head(5)


Product_Name Product_Color  Product_Size
0  SWEATZY_MEN       Crimson            36
1  SWEATZY_MEN       Crimson            37
2  SWEATZY_MEN       Crimson            38
3  SWEATZY_MEN       Crimson            39
4  SWEATZY_MEN       Crimson            40

In [61]:
productsdf.shape

(1170, 3)

Because kids have different shoe sizes than men and women, the possible permutations are calculated separately and appended afterwards.

In [62]:
#This step was done before but repeated for the kids products 
kidssize = []
for i in range(26, 39):
    kidssize.append(i)

kidscategories = [['SWEATZY_KIDS', 'WOVE_KIDS', 'BOOTWEAR_KIDS', 'FROVE_KIDS', 'SWOONIX_KIDS'],
                ['Crimson', 'Navy', 'Olive', 'Baby Blue',
                 'Cyan', 'Coral', 'Azure', 'Denim', 'Peach'],
                kidssize
               ]
print(kidscategories)


[['SWEATZY_KIDS', 'WOVE_KIDS', 'BOOTWEAR_KIDS', 'FROVE_KIDS', 'SWOONIX_KIDS'], ['Crimson', 'Navy', 'Olive', 'Baby Blue', 'Cyan', 'Coral', 'Azure', 'Denim', 'Peach'], [26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38]]


### Listing all permutations for the kids products

In [63]:
import itertools
kidsproducts = list(itertools.product(*kidscategories))


In [64]:

kidsproductsdf = pd.DataFrame (kidsproducts, columns = ['Product_Name', 'Product_Color', 'Product_Size'])
kidsproductsdf.head(5)
kidsproductsdf.shape

(585, 3)

In [65]:
#kidsproductsdf.describe()

### Appending kids dataframe to other dataframe to create one products dataframe

In [66]:
productsdf = productsdf.append(kidsproductsdf)
#productsdf.describe()


### Conditions for product price of different shoes

The price are allocated to a certain shoe based on the product category but also whether it's a kids or adult shoe.

In [67]:
def productprice (row):
    if row['Product_Name'] == 'SWEATZY_KIDS' :
        return 39.95
    if row['Product_Name'] == 'SWEATZY_MEN' or row['Product_Name'] == 'SWEATZY_WOMEN':
        return 49.95
    if row['Product_Name'] == 'WOVE_KIDS' :
        return 49.95
    if row['Product_Name'] == 'WOVE_MEN' or row['Product_Name'] == 'WOVE_WOMEN' :
        return 59.95
    if row['Product_Name'] == 'BOOTWEAR_KIDS' :
        return 64.95
    if row['Product_Name'] == 'BOOTWEAR_MEN' or row['Product_Name'] == 'BOOTWEAR_WOMEN' :
        return 74.95
    if row['Product_Name'] == 'FROVE_KIDS' :
        return 49.95
    if row['Product_Name'] == 'FROVE_MEN' or row['Product_Name'] == 'FROVE_WOMEN' :
        return 59.95
    if row['Product_Name'] == 'SWOONIX_KIDS' :
        return 54.95
    if row['Product_Name'] == 'SWOONIX_MEN' or row['Product_Name'] == 'SWOONIX_WOMEN' :
        return 64.95


In [68]:
productsdf.apply (lambda row: productprice(row), axis=1)

0      49.95
1      49.95
2      49.95
3      49.95
4      49.95
       ...  
580    54.95
581    54.95
582    54.95
583    54.95
584    54.95
Length: 1755, dtype: float64

In [69]:
productsdf['Product_Price'] = productsdf.apply (lambda row: productprice(row), axis=1)
productsdf

Product_Name Product_Color  Product_Size  Product_Price
0     SWEATZY_MEN       Crimson            36          49.95
1     SWEATZY_MEN       Crimson            37          49.95
2     SWEATZY_MEN       Crimson            38          49.95
3     SWEATZY_MEN       Crimson            39          49.95
4     SWEATZY_MEN       Crimson            40          49.95
..            ...           ...           ...            ...
580  SWOONIX_KIDS         Peach            34          54.95
581  SWOONIX_KIDS         Peach            35          54.95
582  SWOONIX_KIDS         Peach            36          54.95
583  SWOONIX_KIDS         Peach            37          54.95
584  SWOONIX_KIDS         Peach            38          54.95

[1755 rows x 4 columns]

The product table has 1755 unique combinations of shoe types, colors, and sizes. 

### Probabilities of shoe sizes

First, a product table was made separately for men, women, and kids shoes. The probability of a product ID being bought is based on a normal distribution of size.

For men, shoe size is normally distributed with an average of 43 and sd = 2.0.

For women, shoe size is normally distributed with an average of 39 and sd = 2.0.

Kids' shoes are not normally distributed considering the difference in height across multiple years. Instead, probabilities were assigned manually.


#### Men

In [70]:
sizes = []
for i in range(35,50):
    sizes.append(i)
print(sizes)

[35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]


In [71]:
import scipy.stats as st

#The normal distribution of shoe sizes
u = 43.00
o = 2.00

#Z-scores are calculated for all possible shoe sizes
Zscores = []
for i in sizes:
    Z = ((i+0.5)-u)/o
    Zscores.append(Z)

print(Zscores)

[-3.75, -3.25, -2.75, -2.25, -1.75, -1.25, -0.75, -0.25, 0.25, 0.75, 1.25, 1.75, 2.25, 2.75, 3.25]


The Z-scores are transformed into the p-values for every shoe size. Note that size 35 was also included even though this is not an actual men shoe size, this was implemented because the probability of shoe size 36 was determined by calculating the P = 35.5 < X < 36.5 

In [72]:
pvalues = []

for i in Zscores:
    p = st.norm.cdf(i)
    pvalues.append(p)
    
print(pvalues)

[8.841728520080377e-05, 0.0005770250423907659, 0.002979763235054555, 0.012224472655044696, 0.040059156863817086, 0.10564977366685535, 0.2266273523768682, 0.4012936743170763, 0.5987063256829237, 0.7733726476231317, 0.8943502263331446, 0.9599408431361829, 0.9877755273449553, 0.9970202367649454, 0.9994229749576092]


In [73]:

sizeprobability = []
for i in range(len(pvalues)):
    p_upper = pvalues[i]
    p_lower = pvalues[i-1]
    p = p_upper - p_lower
    sizeprobability.append(round(p, 4))

del sizeprobability[0]
print(sizeprobability)

[0.0005, 0.0024, 0.0092, 0.0278, 0.0656, 0.121, 0.1747, 0.1974, 0.1747, 0.121, 0.0656, 0.0278, 0.0092, 0.0024]


#### Men shoe size probabilties to dictionary

In [74]:
sizedict = {}
for key in size:
    for prob in sizeprobability:
        sizedict[key] = prob
        sizeprobability.remove(prob)
        break

print(sizedict)

{36: 0.0005, 37: 0.0024, 38: 0.0092, 39: 0.0278, 40: 0.0656, 41: 0.121, 42: 0.1747, 43: 0.1974, 44: 0.1747, 45: 0.121, 46: 0.0656, 47: 0.0278, 48: 0.0092}


The steps explained above are repeated for women's shoes and kids' shoes.

#### Women

In [75]:
womensize = []
for i in range(35,46):
    womensize.append(i)
print(womensize)

wsize = []

for i in range(34, 47):
    wsize.append(i)
print(wsize)

[35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45]
[34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46]


In [76]:
u = 39.00
o = 2.00

Zscores = []
for i in wsize:
    Z = ((i+0.5)-u)/o
    Zscores.append(Z)

print(Zscores)

[-2.25, -1.75, -1.25, -0.75, -0.25, 0.25, 0.75, 1.25, 1.75, 2.25, 2.75, 3.25, 3.75]


In [77]:
pvalues = []

for i in Zscores:
    p = st.norm.cdf(i)
    pvalues.append(p)
    
print(pvalues)

[0.012224472655044696, 0.040059156863817086, 0.10564977366685535, 0.2266273523768682, 0.4012936743170763, 0.5987063256829237, 0.7733726476231317, 0.8943502263331446, 0.9599408431361829, 0.9877755273449553, 0.9970202367649454, 0.9994229749576092, 0.9999115827147992]


In [78]:
wsizeprobability = []
for i in range(len(pvalues)):
    p_upper = pvalues[i]
    p_lower = pvalues[i-1]
    p = p_upper - p_lower
    wsizeprobability.append(round(p, 4))

del wsizeprobability[0]
print(wsizeprobability)

[0.0278, 0.0656, 0.121, 0.1747, 0.1974, 0.1747, 0.121, 0.0656, 0.0278, 0.0092, 0.0024, 0.0005]


In [79]:
wsizedict = {}
for key in womensize:
    for prob in wsizeprobability:
        wsizedict[key] = prob
        wsizeprobability.remove(prob)
        break

print(wsizedict)

{35: 0.0278, 36: 0.0656, 37: 0.121, 38: 0.1747, 39: 0.1974, 40: 0.1747, 41: 0.121, 42: 0.0656, 43: 0.0278, 44: 0.0092, 45: 0.0024}


### Remove women shoes with size > 45 and men shoes < 39

Due to extremely low probabilities, several shoe sizes are removed for men and women to reduce the number of different products.

In [80]:
productsdf = productsdf.drop(productsdf[(productsdf.Product_Size > 45) & (productsdf.Product_Name.str.strip().str[-5:] == 'WOMEN')].index)
productsdf = productsdf.drop(productsdf[(productsdf.Product_Size < 39) & (productsdf.Product_Name.str.strip().str[-4:] == '_MEN')].index)


In [81]:
print(productsdf.iloc[:,0])#.strip()[-5:])

3       SWEATZY_MEN
4       SWEATZY_MEN
5       SWEATZY_MEN
6       SWEATZY_MEN
7       SWEATZY_MEN
           ...     
580    SWOONIX_KIDS
581    SWOONIX_KIDS
582    SWOONIX_KIDS
583    SWOONIX_KIDS
584    SWOONIX_KIDS
Name: Product_Name, Length: 1350, dtype: object


#### Kids

In [82]:
kidssizedict = {26:0.069,
               27:0.069,
               28:0.070,
               29:0.071,
               30:0.074,
               31:0.075,
               32:0.076,
               33:0.078,
               34:0.080,
               35:0.082,
               36:0.084,
               37:0.085,
               38:0.087}

### Function to apply probabilities to Women and Men shoes

In [83]:
def sizeprobability(sizep, productname):
    if productname.strip()[-4:] == '_MEN':
        return sizedict[sizep]
    if productname.strip()[-5:] == 'WOMEN':
        return wsizedict[sizep]
    if productname.strip()[-4:] == 'KIDS':
         return kidssizedict[sizep]

In [84]:
kidssizedict[28]

# print(sizeprobability(36, 'FROVE_KIDS'))

0.07

In [85]:
sizeprobability(26, 'FROVE_KIDS')

0.069

In [86]:
productsdf.head()

Product_Name Product_Color  Product_Size  Product_Price
3  SWEATZY_MEN       Crimson            39          49.95
4  SWEATZY_MEN       Crimson            40          49.95
5  SWEATZY_MEN       Crimson            41          49.95
6  SWEATZY_MEN       Crimson            42          49.95
7  SWEATZY_MEN       Crimson            43          49.95

In [87]:
productsdf['Probability'] = productsdf.apply(lambda x: sizeprobability(x['Product_Size'], x['Product_Name']), axis=1)


In [88]:
#kidsproductsdf.describe()

In [89]:
#productsdf.describe()

In [90]:
display(productsdf)

Product_Name Product_Color  Product_Size  Product_Price  Probability
3     SWEATZY_MEN       Crimson            39          49.95       0.0278
4     SWEATZY_MEN       Crimson            40          49.95       0.0656
5     SWEATZY_MEN       Crimson            41          49.95       0.1210
6     SWEATZY_MEN       Crimson            42          49.95       0.1747
7     SWEATZY_MEN       Crimson            43          49.95       0.1974
..            ...           ...           ...            ...          ...
580  SWOONIX_KIDS         Peach            34          54.95       0.0800
581  SWOONIX_KIDS         Peach            35          54.95       0.0820
582  SWOONIX_KIDS         Peach            36          54.95       0.0840
583  SWOONIX_KIDS         Peach            37          54.95       0.0850
584  SWOONIX_KIDS         Peach            38          54.95       0.0870

[1350 rows x 5 columns]

In [91]:
# Add unique productID 
productsdf1 = productsdf.copy()
productsdf1.insert(0, 'productID', ['P' + str(i) for i in range(int(1e4), int(1e4) + len(productsdf))])
#productsdf1['productID'] = 

In [92]:
# Add column with date a shoe is introduced
import datetime
productsdf1['Date_Introduced'] = 2003
productsdf1.loc[productsdf1.Product_Name.astype(str).str[:7] == 'SWEATZY', "Date_Introduced"] = 2007
productsdf1.loc[productsdf1.Product_Name.astype(str).str[:8] == 'BOOTWEAR', "Date_Introduced"] = 2014

In [93]:
productsdf1.dtypes

productID           object
Product_Name        object
Product_Color       object
Product_Size         int64
Product_Price      float64
Probability        float64
Date_Introduced      int64
dtype: object

In [94]:
# Add different (arbitrary) probabilities for the type of shoe, so that the 'cheap' shoe sells more often
# Multiply this with the existing probabilities to get a new probability 

type_prob = {'SWEATZY': 30,
    'WOVE': 23,
    'BOOTWEAR': 15,
    'FROVE': 20,
    'SWOONIX': 12}

productsdf1['Probability'] = productsdf1['Product_Name'].str.split('_').str[0].map(type_prob)*productsdf1.Probability

In [99]:
import os
os.getcwd()

'/Users/jolyndevis/Documents/SneakerFaqtory/Sneaker-Faqtory'

In [101]:
productsdf1.to_csv(os.getcwd() + '/products.txt', sep='|', index=False)

In [76]:
(productsdf1.Probability/productsdf1.Probability.sum()).sum()

#weights = stacked / stacked.sum()

1.0

In [77]:
?pd.to_csv()

Object `pd.to_csv()` not found.


### Changing index start from 0 to 1

In [78]:
# productsdf.index += 1

In [79]:
productsdf.to_csv('Products.csv')